README.md
[![Build Status](https://travis-ci.org/mmi-cookbooks/metrics-repose.svg)](https://travis-ci.org/mmi-cookbooks/metrics-repose)
[![Code Climate](https://codeclimate.com/github/mmi-cookbooks/metrics-repose/badges/gpa.svg)](https://codeclimate.com/github/mmi-cookbooks/metrics-repose)
[![Coverage Status](https://coveralls.io/repos/mmi-cookbooks/metrics-repose/badge.svg?branch=master&service=github)](https://coveralls.io/github/mmi-cookbooks/metrics-repose?branch=master)

# metrics-repose

This cookbook wraps the standard [repose cookbook](https://github.com/rackerlabs/cookbook-repose) and applies Rackspace Metrics particular settings and filters for query and ingest nodes.  Metrics has two separate repose configurations for query and ingest nodes.

To use this cookbook you'll need to add it via [berkshelf](http://berkshelf.com/) or otherwise get it into your cookbooks dir.  You'll also need to reference the base repose cookbook.

## Supported Platforms

Other platforms are untested.

- Ubuntu 14.04

## Attributes

See the upstream repose cookbook for reference on available attributes (there are a lot).   If setting attributes to override the upstream cookbook, they are set in either attributes/default.rb, recipes/ingest.rb or recipes/query.rb.

## Usage

Include `metrics-repose` in your node's `run_list`:

### metrics-repose::query

```json
{
  "run_list": [
    "recipe[metrics-repose::query]"
  ]
}
```

### metrics-repose::ingest

```json
{
  "run_list": [
    "recipe[metrics-repose::ingest]"
  ]
}
```

### metrics-repose::default

```json
{
  "run_list": [
    "recipe[metrics-repose::default]"
  ]
}
```

## Releasing
To release a new version of this cookbook, do the following:

1. update the CHANGELOG.md file with the new version and changes
2. update the metadata.rb file
3. commit all changes to master (or merge PR, etc.)
4. tag the repo with the matching version for this release

Then you probably want to update some Berksfiles :smile:

## Building and Testing

### Requirements
1. Vagrant
Download it from: https://www.vagrantup.com/downloads.html
2. Virtual Box
Download it from: https://www.virtualbox.org/wiki/Downloads

Optionally for OSX, you can use homebrew install to vagrant and virtualbox together. 

- http://brew.sh/index.html
- http://sourabhbajaj.com/mac-setup/Vagrant/README.html

```
brew cask install virtualbox
brew cask install vagrant
```

### Unit tests
These can be run using the command $ chef exec rspec test/unit  
or with the Rake task $ rake unit

Unit tests will automatically be run in TravisCI on PRs and branch uploads.

### Kitchen and Travis
You can run 'kitchen test' or 'kitchen converge [ingest|query]' to test the cookbook.  This cookbook will run some lint checking on TravisCI when pushed to github.

```
# useful kitchen commands
kitchen create 		# builds the basic VM
kitchen converge 	# tests chef convergence
kitchen verify 		# runs integration tests
kitchen destroy 	# clean everything up
kitchen test 		# basically runs through those 4 steps above) - but does a destroy at the end
kitchen login 		# will log you into the VM in any of the first 4 steps

# useful vagrant commands
vagrant global-status 			# list VMs info and states
vagrant ssh <machine_id> 		# ssh to the vm
vagrant ssh <machine_id> -c "sudo cat /some_file" > some_file # cat a file on the vm to a local file
vagrant halt <machine_id> 		# stop the vm
vagrant destroy <machine_id> 	# destroy the vm, but you should preferably use `kitchen destroy` if possible
```


CHANGELOG.md
# 0.1.22
- add more CORS resp header to merge-header configuration
- only install cors and merge-header on query nodes

# 0.1.21
- add merge-header and cors filters

# 0.1.20
- fix ip_identity white_list_ip_addresses to allow for different ingest and query settings and be overridden in environments

# 0.1.19
- Add GE special rate limit

# 0.1.18
- Change Repose http-conn-pool max to 400

# 0.1.17
- Adding pre-stop and post-start upstart blocks
	
# 0.1.16
- Add tenant handling to repose keystone_v2.cfg.xml to require tenant auth for query nodes and tenanted metrics.

# 0.1.15
- Fix rate limit values set incorrectly in 0.1.12

# 0.1.14
- Add missing crucial line that actually enables add-header filter

# 0.1.13
- Add the add-header filter to inject Content-Type header if it's not present

# 0.1.12
- Tweak rate limit values in repose for ingest and query nodes.
- Set special ingest rate limit for maas-prod tenant, and special query rate limit for lbaas-prod tenant.

# 0.1.11
- Updating wadls to match the correct RBAC roles.  Enabling rax:roles for validation.

# 0.1.10 
- add perf02 to list of environments for repose data_bag credentials

# 0.1.9
- set additivity to false for blueflood-ingest and blueflood-query http logger, this should stop double logging into /var/log/repose/current.log of http logs

# 0.1.8
- Adding tenanted-url to the keystone-v2 filter

# 0.1.7
- Update default number of log files to keep for /var/log/repose/blueflood-<instance>.log, downsize to 6 so we don't cause disk space alerts

# 0.1.6
- Update upstart script to refer to default-java instead of specific version
- Update customer rate-limit defaults to 2000/MIN for ingest and query (was ingest 1000/MIN and query 1500/MIN)

# 0.1.5
- Add the overlimit_429_responsecode attribute (set to true) so that our rate-limiting filter returns 429 status codes when we rate-limit (and not 413)

# 0.1.4
- Adjust log4j2 appender and upstart script so that logs don't fill up disk

# 0.1.3
- Revert back to use the ip-identity filter until we can upgrade to Repose 8 and use ip-user (it doesn't read from X-Forwarded-For in repose 7 filter)
- Add a rate limit group for maas-prod to give them a very large ceiling.

# 0.1.2
- Update global limit for rate-limiting filter
- Fix incorrect attribute names for http_connection_pool and update values 

# 0.1.1
- Update the logging format to add in additional HTTP headers
- Include install recipe in query and ingest setup so that on initial installs, chef doesn't bail when /etc/repose isn't extant

# 0.1.0
- Initial release of metrics-repose


test/unit/spec/query_spec.rb
require 'chefspec'
require_relative 'spec_helper'

describe 'metrics-repose::query' do
  before { stub_resources }
  let(:chef_run) do
    ChefSpec::SoloRunner.new(UBUNTU_OPTS) do |node|
      node.default['blueflood']['query_servers'] = ['127.0.0.1']
    end.converge(described_recipe)
  end

  it 'include recipe repose::install' do
    expect(chef_run).to include_recipe('repose::install')
  end

  it 'create cookbook file /etc/repose/blueflood-query.wadl' do
    expect(chef_run).to create_cookbook_file('/etc/repose/blueflood-query.wadl')
  end
  it 'create file /var/log/repose/blueflood-query.log' do
    expect(chef_run).to create_file('/var/log/repose/blueflood-query.log')
  end
end


test/unit/spec/ingest_spec.rb
require 'chefspec'
require_relative 'spec_helper'

describe 'metrics-repose::ingest' do
  before { stub_resources }
  let(:chef_run) do
    ChefSpec::SoloRunner.new(UBUNTU_OPTS) do |node|
      node.default['blueflood']['ingest_servers'] = ['127.0.0.1']
    end.converge(described_recipe)
  end

  it 'include recipe repose::install' do
    expect(chef_run).to include_recipe('repose::install')
  end

  it 'create cookbook file /etc/repose/blueflood-ingest.wadl' do
    expect(chef_run).to create_cookbook_file('/etc/repose/blueflood-ingest.wadl')
  end
  it 'create file /var/log/repose/blueflood-ingest.log' do
    expect(chef_run).to create_file('/var/log/repose/blueflood-ingest.log')
  end
end


test/unit/spec/spec_helper.rb
# encoding: UTF-8
# -*- mode: ruby -*-
# vi: set ft=ruby :

require 'rspec/expectations'
require 'chefspec'
require 'chefspec/berkshelf'
require 'chef/application'

::UBUNTU_OPTS = {
  platform: 'ubuntu',
  version: '14.04'
}.freeze

def stub_resources; end

at_exit { ChefSpec::Coverage.report! }


recipes/ingest.rb
node.default['repose']['cluster_ids'] = ['blueflood-ingest']
node.default['repose']['content_body_read_limit'] = 1_048_576

include_recipe 'repose::install'

cookbook_file '/etc/repose/blueflood-ingest.wadl'

if %w(stage prod perf01 perf02 qe01 qe02).any? { |e| e.include?(node.chef_environment) }
  credentials = Chef::EncryptedDataBagItem.load('blueflood', "repose_#{node.chef_environment}")
  node.default['repose']['keystone_v2']['username_admin'] = credentials['username']
  node.default['repose']['keystone_v2']['password_admin'] = credentials['password']
end

repose_peers = []
node['blueflood']['ingest_servers'].each do |server|
  repose_peers.push('cluster_id' => node['repose']['cluster_ids'].first,
                    'id' => server.to_s,
                    'hostname' => server.to_s,
                    'port' => node['repose']['ingest']['container_port'])
end
node.default['repose']['peers'] = repose_peers

file '/var/log/repose/blueflood-ingest.log' do
  user 'root'
  group 'root'
  mode 0o0644
end
node.default['repose']['appenders'] = ['<RollingFile name="blueflood-ingest" fileName="/var/log/repose/blueflood-ingest.log"
                     filePattern="/var/log/repose/blueflood-ingest-%i.log">
            <PatternLayout pattern="%m%n"/>
            <Policies>
                <SizeBasedTriggeringPolicy size="1024 MB"/>
            </Policies>
            <DefaultRolloverStrategy max="6"/>
        </RollingFile>']
node.default['repose']['loggers'] = ['<Logger name="ingest-repose-http-log" level="info" additivity="false">
            <AppenderRef ref="blueflood-ingest"/>
        </Logger>']

node.default['repose']['endpoints'] = [
  {
    'cluster_id' => 'blueflood-ingest',
    'id' => 'blueflood-ingest',
    'protocol' => 'http',
    'hostname' => 'localhost',
    'root_path' => '',
    'port' => 2440,
    'default' => true
  }
]
node.default['repose']['slf4j_http_logging']['id'] = 'ingest-repose-http-log'
node.default['repose']['keystone_v2']['groups_in_header'] = true
node.default['repose']['keystone_v2']['uri'] = 'https://identity.api.rackspacecloud.com'
node.default['repose']['keystone_v2']['cache'] = {
  'timeout_variability' => 10,
  'token_timeout' => 600,
  'group_timeout' => 600,
  'endpoints_timeout' => 600
}
node.default['repose']['http_connection_pool']['chunked_encoding'] = false
node.default['repose']['dist_datastore']['port'] = 9002
node.default['repose']['dist_datastore']['cluster_id'] = ['blueflood-ingest']
node.default['repose']['api_validator']['cluster_id'] = ['all']
node.default['repose']['api_validator']['wadl'] = 'blueflood-ingest.wadl'
node.default['repose']['api_validator']['dot_output'] = '/tmp/blueflood-ingest.wadl.dot'
node.default['repose']['api_validator']['enable_rax_roles'] = true
node.default['repose']['api_validator']['check_well_formed'] = false
node.default['repose']['api_validator']['check_grammars'] = true
node.default['repose']['api_validator']['check_elements'] = true
node.default['repose']['api_validator']['check_plain_params'] = true
node.default['repose']['api_validator']['do_xsd_grammar_transform'] = true
node.default['repose']['api_validator']['enable_pre_process_extension'] = true
node.default['repose']['api_validator']['remove_dups'] = true
node.default['repose']['api_validator']['xpath_version'] = '2'
node.default['repose']['api_validator']['xsl_engine'] = 'XalanC'
node.default['repose']['api_validator']['join_xpath_checks'] = false

node.default['repose']['rate_limiting']['cluster_id'] = ['all']
node.default['repose']['rate_limiting']['uri_regex'] = '/v[0-9.]+/(hybrid:)?[0-9]+/limits/?'
node.default['repose']['rate_limiting']['include_absolute_limits'] = true
node.default['repose']['rate_limiting']['use_capture_groups'] = false
node.default['repose']['rate_limiting']['overlimit_429_responsecode'] = true
node.default['repose']['rate_limiting']['global_limits'] = [
  {
    'id' => 'global',
    'uri' => '*',
    'uri-regex' => '.*',
    'value' => 100_000,
    'http-methods' => 'ALL',
    'unit' => 'MINUTE'
  }
]
node.default['repose']['rate_limiting']['limit_groups'] = [
  {
    'id' => 'match-all',
    'groups' => 'IP_Standard',
    'default' => true,
    'limits' => [
      {
        'id' => 'version-tenantId',
        'uri' => '/version/tenantId/*',
        'uri_regex' => '/v[0-9.]+/((hybrid:)?[0-9]+)/.+',
        'http_methods' => 'ALL',
        'unit' => 'MINUTE',
        'value' => 5000
      }
    ]
  },
  {
    'id' => 'maas-prod',
    'groups' => 'IP_Super',
    'default' => false,
    'limits' => [
      {
        'id' => 'near-unlimited',
        'uri' => '/version/tenantId/*',
        'uri_regex' => '/v[0-9.]+/((hybrid:)?[0-9]+)/.+',
        'http_methods' => 'ALL',
        'unit' => 'MINUTE',
        'value' => 100_000
      }
    ]
  }
]

node.default['repose']['keystone_v2']['tenant_handling'] = {
  'validate_tenant' => {
    'url_extraction_regex' => '/v2.0/([^/]+)/.+'
  }
}
node.default['repose']['keystone_v2']['white_list'] = ['/v2.0/?']
node.default['repose']['ip_identity']['white_list_ip_addresses'] = node['repose']['ingest']['ip_identity']['white_list_ip_addresses']


attributes/default.rb
default['repose']['version'] = '7.3.5.0'
default['repose']['connection_timeout'] = 30_000 # in millis
default['repose']['read_timeout'] = 30_000 # in millis
default['repose']['client_request_logging'] = true
default['repose']['jvm_minimum_heap_size'] = '4g'
default['repose']['jvm_maximum_heap_size'] = '4g'
default['repose']['jvm_maximum_file_descriptors'] = '65535'

default['repose']['ingest']['container_port'] = 9001
default['repose']['query']['container_port'] = 9011
default['repose']['graphite_port'] = '2003'

default['repose']['slf4j_http_logging']['format'] = '<![CDATA[
            {"received": "%t", "duration": "%T", "method": "%m", "url": "%U", "status": "%s", "remote_ip": "%a", "local_ip": "%A", "server_port": "%p", "request.accept": "%{accept}i", "request.accept-encoding": "%{accept-encoding}i", "request.Connection": "%{Connection}i", "request.Content-Encoding": "%{Content-Encoding}i", "request.Content-Length": "%{Content-Length}i", "request.Content-Type": "%{Content-Type}i", "request.Method": "%{method}i", "request.Host": "%{Host}i", "request.query_string": "%q", "request.Received": "%{Received}i", "request.SSLClientCertStatus": "%{SSLClientCertStatus}i", "request.SSLClientCipher": "%{SSLClientCipher}i", "request.SSLSessionID": "%{SSLSessionID}i", "request.trans-id": "%{trans-id}i", "request.URL": "%U", "request.User-Agent": "%{user-agent}i", "request.X-Cluster-Client-Ip": "%{X-Cluster-Client-Ip}i", "request.X-Forwarded-For": "%{X-Forwarded-For}i", "request.X-Forwarded-Port": "%{X-Forwarded-Port}i", "request.X-Forwarded-Proto": "%{X-Forwarded-Proto}i", "request.via": "%{via}i", "request.x-authorization": "%{x-authorization}i", "request.x-default-region": "%{x-default-region}i", "request.x-pp-groups": "%{x-pp-groups}i", "request.x-pp-user": "%{x-pp-user}i", "request.x-real-ip": "%{x-real-ip}i", "request.x-roles": "%{x-roles}i", "request.x-tenant-id": "%{x-tenant-id}i", "request.x-tenant-name": "%{x-tenant-name}i", "request.x-user-id": "%{x-user-id}i", "request.x-user-name": "%{x-user-name}i", "response.Content-Length": "%{content-length}o", "response.Content-Type": "%{content-type}o", "response.Date": "%{date}o", "response.server": "%{server}o"}
            ]]>'

default['repose']['filters'] = %w(
  slf4j-http-logging
  header-normalization
  add-header
  keystone-v2
  ip-identity
  rate-limiting
  api-validator
)
default['repose']['services'] = %w(
  dist-datastore
  http-connection-pool
)

default['repose']['http_connection_pool']['chunked_encoding'] = false
default['repose']['http_connection_pool']['socket_timeout'] = 600_000 # in millis
default['repose']['http_connection_pool']['connection_timeout'] = 30_000 # in millis
default['repose']['http_connection_pool']['max_total'] = 400
default['repose']['http_connection_pool']['max_per_route'] = 400

default['repose']['header_normalization']['cluster_id'] = ['all']
default['repose']['header_normalization']['uri_regex'] = nil
default['repose']['header_normalization']['whitelist'] = []
default['repose']['header_normalization']['blacklist'] = [{
  id: 'authorization',
  http_methods: 'ALL',
  headers: %w(
    X-Authorization
    X-Token-Expires
    X-Identity-Status
    X-Impersonator-Id
    X-Impersonator-Name
    X-Impersonator-Roles
    X-Roles
    X-Contact-Id
    X-Device-Id
    X-User-Id
    X-User-Name
    X-PP-User
    X-PP-Groups
    X-Catalog
    X-Subject-Token
    X-Subject-Name
    X-Subject-ID
  )
}]

default['repose']['add_header']['cluster_id'] = ['all']
default['repose']['add_header']['requests'] = [{
  name:      'content-type',
  quality:   '1',
  overwrite: 'false',
  value:     'application/json'
}]

default['repose']['add_header']['responses'] = [{
  name:      'content-type',
  quality:   '1',
  overwrite: 'false',
  value:     'application/json'
}]

default['repose']['ip_user']['cluster_id'] = ['all']

default['repose']['ip_identity']['cluster_id'] = ['all']
default['repose']['ip_identity']['quality'] = 0.2
default['repose']['ip_identity']['white_list_quality'] = 1.0
default['repose']['ingest']['ip_identity']['white_list_ip_addresses'] = ['127.0.0.1']
default['repose']['query']['ip_identity']['white_list_ip_addresses'] = ['127.0.0.1']

default['repose']['upstart']['pre-stop'] = nil
default['repose']['upstart']['post-start'] = nil


recipes/query.rb
node.default['repose']['cluster_ids'] = ['blueflood-query']
node.default['repose']['content_body_read_limit'] = 32768

include_recipe 'repose::install'

cookbook_file '/etc/repose/blueflood-query.wadl'

if %w(stage prod perf01 perf02 qe01 qe02).any? { |e| e.include?(node.chef_environment) }
  credentials = Chef::EncryptedDataBagItem.load('blueflood', "repose_#{node.chef_environment}")
  node.default['repose']['keystone_v2']['username_admin'] = credentials['username']
  node.default['repose']['keystone_v2']['password_admin'] = credentials['password']
end

repose_peers = []
node['blueflood']['query_servers'].each do |server|
  repose_peers.push('cluster_id' => node['repose']['cluster_ids'].first,
                    'id' => server.to_s,
                    'hostname' => server.to_s,
                    'port' => node['repose']['query']['container_port'])
end
node.default['repose']['peers'] = repose_peers

file '/var/log/repose/blueflood-query.log' do
  user 'root'
  group 'root'
  mode 0o0644
end
node.default['repose']['appenders'] = ['<RollingFile name="blueflood-query" fileName="/var/log/repose/blueflood-query.log"
                     filePattern="/var/log/repose/blueflood-query-%i.log">
            <PatternLayout pattern="%m%n"/>
            <Policies>
                <SizeBasedTriggeringPolicy size="1024 MB"/>
            </Policies>
            <DefaultRolloverStrategy max="6"/>
        </RollingFile>']
node.default['repose']['loggers'] = ['<Logger name="query-repose-http-log" level="info" additivity="false">
            <AppenderRef ref="blueflood-query"/>
        </Logger>']

node.default['repose']['endpoints'] = [{
  'cluster_id' => 'blueflood-query',
  'id' => 'blueflood-query',
  'protocol' => 'http',
  'hostname' => 'localhost',
  'root_path' => '',
  'port' => 2500,
  'default' => true
}]

node.default['repose']['filters'] = %w(
  slf4j-http-logging
  merge-header
  cors
  header-normalization
  add-header
  keystone-v2
  ip-identity
  rate-limiting
  api-validator
)

node.default['repose']['slf4j_http_logging']['id'] = 'query-repose-http-log'

node.default['repose']['merge_header']['cluster_id'] = ['all']
node.default['repose']['merge_header']['response_headers'] = %w(
  Access-Control-Allow-Credentials
  Access-Control-Allow-Origin
  Access-Control-Allow-Headers
  Access-Control-Allow-Methods
  Access-Control-Expose-Headers
  Vary
)

node.default['repose']['cors']['cluster_id'] = ['all']
node.default['repose']['cors']['allowed_methods'] = %w(
  GET
)
node.default['repose']['cors']['resources'] = [
  { 'path'            => '/v2.0/.*/views.*',
    'allowed_methods' => %w(
      POST
    ) }
]

node.default['repose']['keystone_v2']['groups_in_header'] = true
node.default['repose']['keystone_v2']['uri'] = 'https://identity.api.rackspacecloud.com'
node.default['repose']['keystone_v2']['cache'] = {
  'timeout_variability' => 10,
  'token_timeout' => 600,
  'group_timeout' => 600,
  'endpoints_timeout' => 600
}
node.default['repose']['tenant_handling']['send_all_tenant_ids'] = true
node.default['repose']['tenant_handling']['validate_tenant'] = {
  'url_extraction_regex' => '.*'
}
node.default['repose']['tenant_handling']['send_tenant_id_quality'] = {
  'default' => '1.0',
  'roles' => '0.5',
  'uri' => '0.5'
}
node.default['repose']['http_connection_pool']['chunked_encoding'] = false
node.default['repose']['dist_datastore']['port'] = 9012
node.default['repose']['dist_datastore']['cluster_id'] = ['blueflood-query']
node.default['repose']['api_validator']['wadl'] = 'blueflood-query.wadl'
node.default['repose']['api_validator']['dot_output'] = '/tmp/blueflood-query.wadl.dot'
node.default['repose']['api_validator']['enable_rax_roles'] = true
node.default['repose']['api_validator']['check_well_formed'] = false
node.default['repose']['api_validator']['check_grammars'] = true
node.default['repose']['api_validator']['check_elements'] = true
node.default['repose']['api_validator']['check_plain_params'] = true
node.default['repose']['api_validator']['do_xsd_grammar_transform'] = true
node.default['repose']['api_validator']['enable_pre_process_extension'] = true
node.default['repose']['api_validator']['remove_dups'] = true
node.default['repose']['api_validator']['xpath_version'] = '2'
node.default['repose']['api_validator']['xsl_engine'] = 'XalanC'
node.default['repose']['api_validator']['join_xpath_checks'] = false

node.default['repose']['rate_limiting']['cluster_id'] = ['all']
node.default['repose']['rate_limiting']['uri_regex'] = '/v[0-9.]+/(hybrid:)?[0-9]+/limits/?'
node.default['repose']['rate_limiting']['include_absolute_limits'] = true
node.default['repose']['rate_limiting']['use_capture_groups'] = false
node.default['repose']['rate_limiting']['overlimit_429_responsecode'] = true
node.default['repose']['rate_limiting']['global_limits'] = [
  {
    'id' => 'global',
    'uri' => '*',
    'uri-regex' => '.*',
    'value' => 10_000,
    'http-methods' => 'ALL',
    'unit' => 'MINUTE'
  }
]
node.default['repose']['rate_limiting']['limit_groups'] = [
  {
    'id' => 'match-all',
    'groups' => 'IP_Standard',
    'default' => true,
    'limits' => [
      {
        'id' => 'ge-special',
        'uri' => '/version/hybrid:2830618/*',
        'uri_regex' => '/v[0-9.]+/(hybrid:2830618)/.+',
        'http_methods' => 'ALL',
        'unit' => 'MINUTE',
        'value' => 4000
      },
      {
        'id' => 'version-tenantId',
        'uri' => '/version/tenantId/*',
        'uri_regex' => '/v[0-9.]+/(?!hybrid:2830618)((hybrid:)?[0-9]+)/.+',
        'http_methods' => 'ALL',
        'unit' => 'MINUTE',
        'value' => 2000
      }
    ]
  }
]

node.default['repose']['keystone_v2']['tenant_handling'] = {
  'validate_tenant' => {
    'url_extraction_regex' => '/v2.0/([^/]+)/.+'
  }
}
node.default['repose']['keystone_v2']['white_list'] = ['/v2.0/?']
node.default['repose']['ip_identity']['white_list_ip_addresses'] = node['repose']['query']['ip_identity']['white_list_ip_addresses']
